lib/puppet/type/cert_key_bundle.rb

```diff
@@ -0,0 +1,96 @@
+Puppet::Type.newtype(:cert_key_bundle) do
+  desc 'combines a certificate and private key into a single file'
+
+  ensurable
+
+  newparam(:path, :namevar => true) do
+    desc "Path to the file that will contain the certificate and private key bundle"
+    isrequired
+  end
+
+  newparam(:certificate) do
+    desc "Path to certificate file to include in bundle"
+    isrequired
+
+    def insync?(is)
+      File.read(self[:path]).include?(File.read(should))
+    end
```

Comment:
As currently written, this would pass if there is some extra content included in the bundle which was not defined by Puppet. I'm unsure how much of a concern that really is, but one thought is what if extra certificates were defined that shouldn't be there? In that case, puppet would still treat the resource as being in sync.

Is there any way to test that the bundle itself is in sync rather than testing the key and cert separately?

I thought about one approach which would not really work. That would be something like this for the certificate:

```ruby
    def insync?(is?)
      @cert_is = is
      @cert_should = should
      if defined?(@key_is) and defined?(@key_should)
        bundle_insync?(@cert_is, @cert_should, @key_is, @key_should)
      else
        false
      end
    end
```

This for the key:

```ruby
    def insync?(is?)
      @key_is = is
      @key_should = should
      @key_should = format_as_pkcs_1(@key_should) if self[:force_pkcs_1]
      if defined?(@cert_is) and defined?(@cert_should)
        bundle_insync?(@cert_is, @cert_should, @key_is, @key_should)
      else
        false
      end
    end
```

And the private method:

```ruby
  def cert_and_key_insync?(cert_is, cert_should, key_is, key_should)
    bundle_is = [key_is, cert_is].join("\n")
    bundle_should = [key_should, cert_should].join("\n")
    
    bundle_is == bundle_should
  end
```

But this has a number of pitfalls that I can think of:

1. It seems like abuse of instance variables and you could definitely get into a broken state if the key was previously incorrect, has now been corrected, but the value for `@key_is` has not yet updated and we call `insync?` within the certificate parameter scope first -- I assume that would be possible if Puppet regenerates the resource and then checks again if it is in sync.

2. Also you could get incorrect logs regarding the source of the failure, if the method that checks whether the certificate is in sync returns false because the key was wrong.

So clearly that approach is unsatisfactory.

Reply:
A different idea is to use regular expressions w/ selection groups, so that e.g. the `insync?` method for the private key could return true if the bundle matches a regular expression for a key + cert bundle, and the selection for the key matches exactly what it should be.

Likewise the `insync?` method for the cert would return true if the bundle matches a regular expression for a key + cert bundle, and the selection for the cert matches exactly what it should be.

Reply:
I dropped the `insync?` as that appears to only be defined for propteries and not params. And it's unnecessary when the `exist?` check is ensuring the full formatted content.